contracts/external/nexus/WrappedCoveredPrincipalToken.sol

```diff
@@ -0,0 +1,145 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity 0.8.0;
+
+import { ERC20PermitWithSupply, ERC20Permit } from "../../libraries/ERC20PermitWithSupply.sol";
+import { Authorizable } from "../../libraries/Authorizable.sol";
+import { ITranche } from "../../interfaces/ITranche.sol";
+import { SafeERC20, IERC20 } from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import { IERC20Metadata } from "@openzeppelin/contracts/token/ERC20/extensions/IERC20Metadata.sol";
+import { EnumerableSet } from "@openzeppelin/contracts/utils/structs/EnumerableSet.sol";
+
+contract WrappedCoveredPrincipalToken is ERC20PermitWithSupply, Authorizable {
+    using EnumerableSet for EnumerableSet.AddressSet;
+    using SafeERC20 for IERC20;
+
+    // Address of the base/underlying token which is used to buy the yield bearing token from the wrapped position.
+    // Ex - Dai is used to buy the yvDai yield bearing token
+    address public immutable baseToken;
+
+    // Enumerable address list, It contains the list of allowed tranches that are covered by this contract
+    // Criteria to choose the tranches are -
+    // a). Tranche should have same underlying/base token (i.e ETH, BTC, USDC).
+    // b). Should have the similar risk profiles.
+    EnumerableSet.AddressSet private _allowedTranches;
+
+    // Emitted when new tranche get whitelisted.
+    event TrancheAdded(address _tranche);
+
+    /// @notice Modifier to validate the tranche is whitelisted or not.
+    modifier isValidTranche(address _tranche) {
+        require(isAllowedTranche(_tranche), "WFP:ALREADY_EXISTS");
+        _;
+    }
+
+    ///@notice Initialize the wrapped token.
+    ///@dev    Wrapped token have 18 decimals, It is independent of the baseToken decimals.
+    constructor(address baseToken_, address owner_)
+        ERC20Permit(
+            _processName(IERC20Metadata(baseToken_).symbol()),
+            _processSymbol(IERC20Metadata(baseToken_).symbol())
+        )
+    {
+        baseToken = baseToken_;
+        _authorize(owner_);
+        setOwner(owner_);
+    }
+
+    ///@notice Allows to create the name for the wrapped token.
+    function _processName(string memory _tokenSymbol)
+        internal
+        pure
+        returns (string memory)
+    {
+        return
+            string(
+                abi.encodePacked("Wrapped", _tokenSymbol, "Futures Principal")
+            );
+    }
+
+    ///@notice Allows to create the symbol for the wrapped token.
+    function _processSymbol(string memory _tokenSymbol)
+        internal
+        pure
+        returns (string memory)
+    {
+        return string(abi.encodePacked("ep:", "W", _tokenSymbol));
```

Comment:
I think the `ep` might be redundant. 
what about `ep:WePyvUSDC-29OCT21` -> `WePyvUSDC-29OCT21` ?

I'm also thinking about the possibility of using this contract to wrap principal token derivatives, in which case forcing the `ep` would be annoying